### ISSUE TITLE
Flaky: Http3LoopbackTestSuite handshake times out with local IdleTimeout — load-dependent, likely the same defect as #367

### ISSUE THREAD

**Issue by thebehera** [DitchOoM/socket#450]
`Http3LoopbackTestSuite` fails intermittently with a QUIC handshake that never completes. Two different tests have been observed failing this way, which is why this is filed as one defect rather than two flakes.

## Sightings

**CI, `build-linux / Linux / JVM / JS / Android`** (run 32586338373):

```
JvmHttp3LoopbackTest[jvm] > webTransport_serverOpensUniStream_clientReceives[jvm] FAILED
    kotlinx.coroutines.TimeoutCancellationException at JvmHttp3LoopbackTest.kt:34
```

A re-run of the same job on the same commit passed, so it is not commit-specific.

**Local**, 1 failure in 25 runs of the same class:

```
productionServerRole_dynamicQpackRoundTrip[jvm]
  QuicCloseException: QUIC handshake failed [local: IdleTimeout (0x-1)]
  (session=8f2f64bd91fe71d95e810000b326bfe5296aab78)
```

`local: IdleTimeout` during the handshake — the connection never came up, and our side is the one that gave up.

## Likely the same defect as #367

#367 is `AppleWebTransportTest.connect_peerResetsStream` failing with "QUIC handshake IdleTimeout with an empty server trace". Same phase (handshake), same typed reason (IdleTimeout), same emptiness on the far end. Different suite and platform, which is what has kept them filed apart. Worth resolving together.

## It looks load-dependent, and that matters for reproducing it

| arm | runs | failures |
|---|---|---|
| local, machine also running other Gradle + CI polling | 25 | **1** |
| local, quiet machine (instrumented) | 40 | 0 |
| local, quiet machine (control, uninstrumented) | 40 | 0 |

The only reproduction so far happened while the machine was busy. Both quiet 40-run loops came back clean, on identical code. A handshake that times out only under contention is consistent with CI's loaded shared runners failing more often than a developer laptop.

The two 40-run arms are also the control measurement for the instrumentation added in the linked PR: equal at 0, so the marks are not masking the race (the concern that println probes hid #393 outright).

**So the next step is not another quiet loop.** It is a loop under deliberate CPU load, which is the cheapest way to turn a ~4% flake into something reproducible on demand.

## Why there is no regression test yet

There is no identified defect to pin. A test written now would encode a guess about a bug we have not diagnosed and pass for reasons unrelated to it. The sequence is reproduce → capture → diagnose → *then* pin.

## What has been done

The linked PR makes the next occurrence self-diagnosing:

- the existing diagnostics report (last step, typed QUIC reason, both ends' traces) is now folded into the thrown exception, not only system-out — on CI the report existed but was unreadable, because the job log shows the exception and the test XML is clobbered by the JNI→FFM re-run;
- the WebTransport test gained stage marks on both ends. It had none, and three of its stages were the same `withTimeout(5.seconds)`, so nothing could say which stalled.

The next red run should name the stage and carry both traces.

**Comment by thebehera** [comment]
**Deterministic reproducer found — seed 4.**

`HandshakeIdleTimeoutSimTests` (PR #481) runs a real quiche client against a real quiche server over a seeded `ImpairedPipe`, with quiche's own timers on a virtual clock. At `loss = 0.30`, 12 seeds:

```
[#450 seed search] loss=0.30, virtual clock, 12 seeds
  established     : 11 [1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12]
  LOCAL IDLE T/O  : 1 [4]   <-- the #450 signature
```

Seed 4 produces `Closed(ByLocal(IdleTimeout))` during the handshake — the local sighting in this issue's body — and does so **5/5 identically**, at establishment bounds of 3s, 10s and 30s.

```
./gradlew :socket-quic-quiche:jvmTest \
  --tests 'com.ditchoom.socket.quic.HandshakeIdleTimeoutSimTests' --rerun
```

## What this replaces

The load hunts this issue recommended do not work:

| hunt | result |
|---|---|
| 18-core Mac, 16 burners, load average 30–56, 40 runs | 0/40 |
| 4-vCPU CI runner, 25 runs (`flake-hunt.yaml`) | 0/25 |

A hunt whose control never reproduces cannot evaluate a fix. The seed search does, and yields a case that re-runs on demand under a debugger rather than a rate.

## Correction to an earlier claim on this issue

I briefly filed #480 claiming a stalled handshake never terminates on its idle timeout. **That was wrong** — measured directly, the client closes with `ByLocal(IdleTimeout)` exactly as an established connection does, because quiche arms the idle timer on the first ack-eliciting send (RFC 9000 §10.1). The apparent hang was the simulator's eagerly-accepted server, which is unreachable in production. #480 has been retitled to the small real gap it left behind.

## Mechanism still open

Nothing here yet explains why loopback stalls for ≥10s in the field. On loopback nothing is lost for that long, so the standing suspect remains the driver/reader coroutine being starved while quiche's clock keeps running — the same starvation family as #470 and #478. Seed 4 is now a way to study a stalled handshake deterministically, not yet the field mechanism.

**Comment by thebehera** [comment]
**Root cause found, reproduced 4–5 times in 20 000 handshakes, and fixed. It is not in QUIC at all —
it is in `UdpSocket.bind`.**

## The mechanism

`withQuicServer(port = 0, host = null)` binds through `UdpSocket.bind(null, 0)`. `DatagramChannel.open()`
takes no protocol family, so the JDK returns an `AF_INET6` socket with `IPV6_V6ONLY` cleared and turns
the `0.0.0.0` wildcard into `in6addr_any` — one socket meant to serve both families. **On Darwin that
socket does not own the IPv4 half of its port.** Measured directly, three ways:

| already holding the port | dual-stack `[::]:P` bind | who gets `127.0.0.1:P` | who gets `[::1]:P` |
|---|---|---|---|
| `AF_INET` **wildcard** `0.0.0.0:P` | **succeeds** | the `AF_INET` socket | the dual-stack socket |
| `AF_INET` specific `127.0.0.1:P` | refused (`EADDRINUSE`) | — | — |
| `AF_INET6` with `IPV6_V6ONLY=1` | refused (`EADDRINUSE`) | — | — |

Exactly one shape is dangerous and it is the silent one. `bind(0)` picks the port from the IPv6 table,
so it can hand a server a port whose IPv4 half belongs to an unrelated daemon. `lsof` at the moment of
a captured failure:

```
lsof    homed  1057  50u  IPv4  UDP *:49928     <- macOS daemon, AF_INET wildcard
lsof    java  99281  71u  IPv6  UDP *:49928     <- our server, dual-stack
netstat udp4   0  0  *.49928  *.*
netstat udp46  0  0  *.49928  *.*
```

The server is open, healthy, parked in `select()` — and permanently deaf over IPv4 for its whole life.
macOS's ephemeral range is 16 384 ports and a handful are held this way, so the rate is ~1 in 8 000;
measured at **1 in 4 000** over 40 000 server binds.

## The capture, and how each reading was excluded

```
attempt #1385 serverPort=49928 outcome=LocalClose(reason=ByLocal(error=IdleTimeout))
    client sent 6 datagram(s); server receive loop saw headerInfo=0 accept=0 connRecv=0
    post-mortem probe: sent 32B to 127.0.0.1:49928; headerInfo 0 -> 0
    threads: total=34 inSelect=2 ioParallelism=<default 64>
      [DefaultDispatcher-worker-15] sun.nio.ch.KQueue.poll <- KQueueSelectorImpl.doSelect
        <- SelectorImpl.select <- NioDatagramChannelCore.receive
```

- the client **did** send — six Initials, each recorded only on `SendOutcome.Sent`;
- the server's receive loop called `headerInfo` **zero** times, so nothing was rejected after arrival —
  nothing arrived;
- a probe datagram from a socket unrelated to that connection did not reach it either, so it is not
  per-connection;
- its `select()` was parked and healthy on its own kqueue, so the socket was not broken.

Which leaves one thing: the datagrams were delivered elsewhere. `lsof` said where.

## The load-dependence in this issue's table is a confounder

Three arms, identical except for whether the server binds a fresh ephemeral port each time. `main`,
18-core Mac, 12–16 burners, load average ~20:

| arm | handshakes | `ByLocal(IdleTimeout)` |
|---|---|---|
| sequential, one long-lived server | 20 000 | 0 |
| concurrent ×8, one long-lived server | 20 000 | 0 |
| **fresh server per handshake** | **20 000** | **4** |
| **fresh server per handshake (counted)** | **20 000** | **5** |

Same load, same process, same client dial. The variable is **binds**, not load — and that is why both
hunts this issue recommended came back empty rather than unlucky: 40 runs of a 32-test class is roughly
2 000 server binds, so at 1-in-4 000 the *expected* number of failures was 0.5. The hunts were
under-powered, not unrepresentative. A per-handshake instrument does 2 000 binds in twenty seconds.

The client is immune for a structural reason, which is itself a check on the model: a client socket is
`connect()`ed, so its fully-specified 4-tuple out-ranks any wildcard holder. 20 000 client binds in the
sequential arm produced 0 failures.

## Correction: seed 4 (PR #481) reproduces the *report*, not this defect

This issue records seed 4 as "the #450 signature". Bisected with the RNG taken out of the picture
(`loss = 0` plus an explicit drop script):

| drop script | idle 2s | idle 10s |
|---|---|---|
| nothing | Established, 10 ms | Established, 10 ms |
| the client's first Initial | Established, 1 009 ms | Established, 1 009 ms |
| its first PTO retransmission | Established, 1 029 ms | Established, 1 029 ms |
| **both** | **LocalIdleTimeout, 2 997 ms** | **Established, 3 007 ms** |

Two consecutive losses of one Initial (p = 0.09 at 30 % loss, so 1 seed in 12 is the binomial rate)
inside the 2 s idle budget the *sim* chose. quiche floors the idle timeout at 3×PTO — RFC 9000 §10.1 —
so the deadline is 2 997 ms, and §10.1 restarts the idle timer on a send only when no other
ack-eliciting packet has been sent since the last receive, so the 999 ms retransmission does not push it
out; meanwhile the PTO backoff schedules the third Initial at 999 + 1 998 = **2 997 ms**, the same
instant. It dies one tick before the transmission that would have saved it, exactly as specified.

At the 10 s idle timeout `Http3LoopbackTestSuite` actually configures, the identical drop script
establishes. A loopback path loses nothing for 10 s, so that seed never modelled this. It is now pinned
as an assertion rather than left for the next reader to misread, since the sweep will keep reporting
`LOCAL IDLE T/O: 1 [4]` forever.

## The fix

A wildcard bind takes its port from the **IPv4** table first: `bind(0)` on a throwaway `AF_INET` probe
is handed a port free for IPv4 by definition, the probe is closed (UDP has no `TIME_WAIT`), and the
dual-stack socket takes that port. An explicitly requested port whose IPv4 half is held now fails with
the same `BindException` Linux already raises, instead of succeeding into a deaf socket. Both actuals
that bind a dual-stack wildcard are covered — `UdpSocket.jvm.kt` and `UdpSocket.apple.kt`; the Linux
actual binds an IPv4 wildcard and is not exposed.

Same four arms after the fix: **80 000 handshakes, 0 failures.**

## What this does not cover

The other sighting in this issue's body — the Linux CI `TimeoutCancellationException at
JvmHttp3LoopbackTest.kt:34` — carried no typed reason. Linux refuses the conflicting bind at the kernel,
so it is not exposed to this mechanism, and that run cannot be attributed here. #499 has since made a
stalled establishment report a typed reason, so a recurrence will name itself rather than arriving as a
bare cancellation.

#367 stays open on its own (`Refs`): the Apple actual had the identical exposure and is fixed with its
own red/green/mutation, but an `NWConnection` client that resolves `localhost` to `::1` is immune by the
table above, so whether that sighting was this is not decidable from its report.


Fix + instruments: #535
